### PR TITLE
Adding Coco Detection Dataset

### PR DESCRIPTION
**Pull request by Louis-Dupont**

### CODE REVIEW

File: src/data_gradients/datasets/detection/coco_detection_dataset.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    ```python
+    from data_gradients.datasets.detection import COCODetectionDataset
+
+    train_set = COCODetectionDataset(root_dir="<path/to/dataset_root>", split="train", year=2017)
+    val_set = COCODetectionDataset(root_dir="<path/to/dataset_root>", split="val", year=2017)
+    ```
+    """
+
+    def __init__(self, root_dir: str, split: str, year: Union[int, str] = 2017):
```
Comment (BloodAxe) [inline src/data_gradients/datasets/detection/coco_detection_dataset.py:36]: I don't think the `year` argument here is the right thing to do.
This may be ok for academic COCO dataset, but if someone exports dataset from Roboflow or any other external platform in COCO format, the notion of the year can be easily missing.
However the class constructor expects this input.
I think we should change this API to take only `images_dir` and `annotation_json` and that's it. Even root dir looks redundant to me, tbh.

Comment (BloodAxe) [inline src/data_gradients/datasets/detection/coco_detection_dataset.py:36]: Do we really need a second class here?

File: src/data_gradients/datasets/detection/coco_format_detection_dataset.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    def load_labels(self, index: int) -> np.ndarray:
+        annotations = self.base_dataset[index][1]
+
+        labels = []
+        for annotation in annotations:
+            class_id = annotation["category_id"]
+            x, y, w, h = annotation["bbox"]
+            labels.append((int(class_id), float(x), float(y), float(w), float(h)))
+        return np.array(labels)
```
Comment (BloodAxe) [inline src/data_gradients/datasets/detection/coco_format_detection_dataset.py:93]: You may want to write something like `np.array(labels, dtype=np.float32).reshape(-1, 5)` to ensure even for empty sample the labels you return from this function will be 2-dimensional array. And explicit dtype is also necessary from the same considerations - to guarantee consistency for empty & non-empty samples

File: src/data_gradients/datasets/detection/coco_format_detection_dataset.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    def __len__(self) -> int:
+        return len(self.base_dataset)
+
+    def __getitem__(self, index: int) -> Tuple[np.ndarray, np.ndarray]:
```
Comment (BloodAxe) [inline src/data_gradients/datasets/detection/coco_format_detection_dataset.py:98]: Why do we have load_image & load_labels? 
I see you're calling `self.base_dataset[index]` two times (one for load_image, second for load_labels) which is double work we can easily avoid.

File: tests/unit_tests/datasets/detection/test_coco_detection_dataset.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import os
+import unittest
+
+from torch.utils.data import DataLoader
+
+from data_gradients.managers.detection_manager import DetectionAnalysisManager
+from data_gradients.datasets.detection.coco_detection_dataset import COCODetectionDataset
+
+
+class CocoSegmentationDatasetTest(unittest.TestCase):
```
Comment (BloodAxe) [inline tests/unit_tests/datasets/detection/test_coco_detection_dataset.py:10]: Coco -> COCO

File: src/data_gradients/datasets/detection/coco_detection_dataset.py (hunk 1 above)
Comment (BloodAxe) [reply src/data_gradients/datasets/detection/coco_detection_dataset.py:36]: Ok, I'll elaborate.
What is the added value of this class and does it outweighs the added support cost of the new class, whose only purpose is to support multiple versions of the official COCO datasets?

Comment (Louis-Dupont) [reply src/data_gradients/datasets/detection/coco_detection_dataset.py:36]: PRO
- Ease of use: the user does not need to dig into the dataset folders to know its inner structure.
- (Consequence of previous) Less likely for users to make mistakes, since we handle path mapping

CON
- If someone uses this class on a version of COCO that is structured differently, then the errors will not be explicit (because we assume it is indeed COCO, and structured like COCO)
- (and of course) adding more code

Do you see other points ?
I think the pros>cons, so I would tend to think it's better to keep, but I'm good with either choice. What do you think?

File: src/data_gradients/datasets/detection/coco_format_detection_dataset.py (hunk 3 above)
Comment (Louis-Dupont) [reply src/data_gradients/datasets/detection/coco_format_detection_dataset.py]: changed in f803c2b5e77fa25aa0a82825cad4ecc6c2d5dc31

File: tests/unit_tests/datasets/detection/test_coco_detection_dataset.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    def setUp(self):
+        project_root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(__file__)))))
+        self.train_set = COCODetectionDataset(root_dir=os.path.join(project_root, "example_dataset", "tinycoco"), split="train", year="2017")
+        self.val_set = COCODetectionDataset(root_dir=os.path.join(project_root, "example_dataset", "tinycoco"), split="val", year="2017")
+
+    def test_coco_dataset(self):
+        da = DetectionAnalysisManager(
+            report_title="Testing Data-Gradients NEW",
```
[the remark this answers is not in the document]
Comment (Louis-Dupont) [reply tests/unit_tests/datasets/detection/test_coco_detection_dataset.py]: Changed in af1fdb68e1f3d7f6c4816d106c0c816e99f97483

File: src/data_gradients/datasets/detection/coco_detection_dataset.py (hunk 1 above)
Comment (BloodAxe) [reply src/data_gradients/datasets/detection/coco_detection_dataset.py:36]: As I said - I don't see a big advantage in two classes, but that's not a deal breaker to me

**Review by BloodAxe — APPROVED** [review]
LGTM
